### ISSUE TITLE
[QUESTION] Truemail not working after some time running in rails process

### ISSUE THREAD

**Issue by mtnstar** [truemail-rb/truemail#235]
### New bug checklist

- [x] I have updated `truemail` to the latest version
- [x] I have read the [Contribution Guidelines](https://github.com/truemail-rb/truemail/blob/master/CONTRIBUTING.md)
- [x] I have read the [documentation](https://truemail-rb.org/truemail-gem)
- [x] I have searched for [existing GitHub issues](https://github.com/truemail-rb/truemail/issues)

### Bug description

we're using Truemail for validating e-mail addresses in our opensource community software [hitobito](github.com/hitobito/hitobito)

it works perfectly until some time has passed, then it just returns false for every e-mail checked. we're using it for example in the Person model: https://github.com/hitobito/hitobito/blob/master/app/models/concerns/validated_email.rb#L17

after some time, sometimes days, sometimes only hours, it stops working and Truemail.valid? always returns false. To fix this, we have to restart the rails server, then it works perfectly again for some time. It affects multiple environments running in different openshift projects. 

we created a health check endpoint for monitoring truemail: https://github.com/hitobito/hitobito/blob/8a6ef125be5f67413dec8f0cfc5e2c87dfa0f2ce/app/domain/app_status/truemail.rb#L8

if it's not working anymore, a simple `touch tmp/restart.txt` fixes the problem.

since we do not have any clue what's causing this problems, we're reaching out here. Maybe someone has any idea what could cause this strange problem.

##### Complete output when running truemail, including the stack trace and command used

<details>
  <pre>no stack trace here because it's not throwing any exceptions at all</pre>
</details>

**Comment by bestwebua** [comment]
@mtnstar Hi! Thanks for report. Very strange issue... But you didn't provide any context of truemail. I mean some logs, when this situation happens. Could you configure logging for `truemail` and post results here when it will happen again? Also you can run from rails console `Truemail.validate(some_email)` to receive more context.

<img width="890" alt="Screenshot 2022-11-09 at 10 52 22" src="https://user-images.githubusercontent.com/14002692/200798459-4e01986c-0f7a-49b2-a04f-74a2847bdf9f.png">

**Comment by mtnstar** [comment]
@bestwebua thx for the input, I now enabled the debug log for some environments and hopping to catch some errors the next time Truemail stops working accordingly.

**Comment by carlobeltrame** [comment]
@bestwebua In the meantime, this happens every day on at least some of our ~10-15 production instances. Below is an example log entry, triggered from a health check call we installed:
```
E, [2022-12-02T15:48:02.740667 #1] ERROR -- : Truemail mx validation for hitobito@puzzle.ch failed (mx: target host(s) not found)
CONFIGURATION SETTINGS:
whitelist validation: false
not rfc mx lookup flow: false
smtp fail fast: false
smtp safe check: false
email pattern: default gem value
smtp error body pattern: default gem value
```

But if I then open a new rails console in that same container which continually has the error and execute the command you proposed, I get the following (validation success):
```
[1] pry(main)> Truemail.validate('hitobito@puzzle.ch')
I, [2022-12-02T15:51:29.000450 #18716]  INFO -- : Truemail mx validation for hitobito@puzzle.ch was successful

CONFIGURATION SETTINGS:
whitelist validation: false
not rfc mx lookup flow: false
smtp fail fast: false
smtp safe check: false
email pattern: default gem value
smtp error body pattern: default gem value

=> #<Truemail::Validator:0x00000000022f3940
 @result=
  #<struct Truemail::Validator::Result
   success=true,
   email="hitobito@puzzle.ch",
   domain="puzzle.ch",
   mail_servers=["5.102.145.24", "5.102.146.40"],
   errors={},
   smtp_debug=nil,
   configuration=
    #<Truemail::Configuration:0x00000000022f39b8
     @blacklisted_domains=[],
     @blacklisted_mx_ip_addresses=[],
     @connection_attempts=2,
     @connection_timeout=2,
     @default_validation_type=:mx,
     @dns=[],
     @email_pattern=/(?=\A.{6,255}\z)(\A([\p{L}0-9]+[\w\p{L}.+!~,'&%#*^`{}|\-\/?=$]*)@((?i-mx:[\p{L}0-9]+([\-.]{1}[\p{L}0-9]+)*\.\p{L}{2,63}))\z)/,
     @logger=#<Truemail::Logger:0x00000000049f4210 @event=:all, @file="/opt/app-root/src/log/truemail.log", @stdout=true>,
     @not_rfc_mx_lookup_flow=false,
     @response_timeout=2,
     @smtp_error_body_pattern=/(?=.*550)(?=.*(user|account|customer|mailbox)).*/i,
     @smtp_fail_fast=false,
     @smtp_port=25,
     @smtp_safe_check=false,
     @validation_type_by_domain={},
     @verifier_domain="puzzle.ch",
     @verifier_email="hitobito@puzzle.ch",
     @whitelist_validation=false,
     @whitelisted_domains=[]>>,
 @validation_type=:mx>
```
Meanwhile the rails puma process keeps on failing to validate even the simplest email addresses. So it must be something local to the process which runs Truemail.

Maybe it has something to do with us running this application on an OpenShift (Kubernetes) architecture? The spontaneous way in which this error appears makes me think that Truemail / the MX lookup might try to keep some connection open, which breaks if the running pod is moved to another kubernetes node maybe..?

**Comment by bestwebua** [comment]
Hello, @carlobeltrame! Thanks for your intention to figure out why truemail didn't work as expected in this case.

> Maybe it has something to do with us running this application on an OpenShift (Kubernetes) architecture? The spontaneous way in which this error appears makes me think that Truemail / the MX lookup might try to keep some connection open, which breaks if the running pod is moved to another kubernetes node maybe..?

I was surprised... Take a look for result from dockerized truemail:

```bash
➜ http localhost:9292/version Authorization:token
HTTP/1.1 200 OK
Connection: close
Content-Type: application/json
Server: thin

{
    "core": "3.0.1",
    "platform": "ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [x86_64-linux-musl]",
    "version": "0.5.0"
}

➜ http localhost:9292 email==hitobito@puzzle.ch Authorization:token
HTTP/1.1 200 OK
Connection: close
Content-Type: application/json
Server: thin

{
    "configuration": {
        "blacklisted_domains": null,
        "blacklisted_emails": null,
        "blacklisted_mx_ip_addresses": null,
        "dns": null,
        "email_pattern": "default gem value",
        "not_rfc_mx_lookup_flow": false,
        "smtp_error_body_pattern": "default gem value",
        "smtp_fail_fast": false,
        "smtp_safe_check": false,
        "validation_type_by_domain": null,
        "whitelist_validation": false,
        "whitelisted_domains": null,
        "whitelisted_emails": null
    },
    "date": "2022-12-03 13:24:51 +0000",
    "email": "hitobito@puzzle.ch",
    "errors": null,
    "smtp_debug": null,
    "success": true,
    "validation_type": "mx"
}
```

And running from MacOS 13.0.1 (22A400):

```ruby
[1] pry(main)> require 'truemail'
=> true
[2] pry(main)> Truemail::VERSION
=> "3.0.1"
[3] pry(main)> RUBY_VERSION
=> "3.1.2"
[4] pry(main)> Truemail.configure { |c| c.verifier_email = 'admin@bestweb.com.ua' }
=> #<Truemail::Configuration:0x000000011109fbe8
 @blacklisted_domains=[],
 @blacklisted_emails=[],
 @blacklisted_mx_ip_addresses=[],
 @connection_attempts=2,
 @connection_timeout=2,
 @default_validation_type=:smtp,
 @dns=[],
 @email_pattern=/(?=\A.{6,255}\z)(\A([\p{L}0-9]+[\w\p{L}.+!~,'&%#*^`{}|\-\/?=$]*)@((?i-mx:[\p{L}0-9]+([\-.]{1}[\p{L}0-9]+)*\.\p{L}{2,63}))\z)/,
 @not_rfc_mx_lookup_flow=false,
 @response_timeout=2,
 @smtp_error_body_pattern=/(?=.*550)(?=.*(user|account|customer|mailbox)).*/i,
 @smtp_fail_fast=false,
 @smtp_port=25,
 @smtp_safe_check=false,
 @validation_type_by_domain={},
 @verifier_domain="bestweb.com.ua",
 @verifier_email="admin@bestweb.com.ua",
 @whitelist_validation=false,
 @whitelisted_domains=[],
 @whitelisted_emails=[]>
[5] pry(main)> Truemail.validate('hitobito@puzzle.ch', with: :mx)
=> #<Truemail::Validator:0x000000011132cd48
 @result=
  #<struct Truemail::Validator::Result
   success=false, # WTF?????
   email="hitobito@puzzle.ch",
   domain="puzzle.ch",
   mail_servers=[],
   errors={:mx=>"target host(s) not found"}, # WTF?????
   smtp_debug=nil,
   configuration=
    #<Truemail::Configuration:0x000000011132ce88
     @blacklisted_domains=[],
     @blacklisted_emails=[],
     @blacklisted_mx_ip_addresses=[],
     @connection_attempts=2,
     @connection_timeout=2,
     @default_validation_type=:smtp,
     @dns=[],
     @email_pattern=/(?=\A.{6,255}\z)(\A([\p{L}0-9]+[\w\p{L}.+!~,'&%#*^`{}|\-\/?=$]*)@((?i-mx:[\p{L}0-9]+([\-.]{1}[\p{L}0-9]+)*\.\p{L}{2,63}))\z)/,
     @not_rfc_mx_lookup_flow=false,
     @response_timeout=2,
     @smtp_error_body_pattern=/(?=.*550)(?=.*(user|account|customer|mailbox)).*/i,
     @smtp_fail_fast=false,
     @smtp_port=25,
     @smtp_safe_check=false,
     @validation_type_by_domain={},
     @verifier_domain="bestweb.com.ua",
     @verifier_email="admin@bestweb.com.ua",
     @whitelist_validation=false,
     @whitelisted_domains=[],
     @whitelisted_emails=[]>>,
 @validation_type=:mx>
```

**Comment by carlobeltrame** [comment]
Very strange. hitobito@puzzle.ch is a normal email address which we know exists because we control it. But our problem is not limited to this address, but concerns any email address. Normally Truemail works as expected, but then spontaneously stops working inside our rails puma process. By "stops working" I mean it stops working completely. Checking any email address will take long (~10-30 seconds) and then fail.

So, based on my suspicion of some network problem when something spontaneously changes in a kubernetes cluster, can you point me to the code where you do the actual MX request which fails here? Then we can try to debug deeper and find out whether the problem lies in a dependency of truemail, or within truemail itself.

**Comment by bestwebua** [comment]
@carlobeltrame Thank you for your investigation ❤️ Today I have investigated this issue too, and I think it's 100% network issue. To make sure this is a truth I suggest to add some network status loggers to DNS validation layer.

**Comment by bestwebua** [comment]
You can temporary monkey patch `Truemail::Wrapper#call` to expose network error context:

https://github.com/truemail-rb/truemail/blob/8a570ff47e4d6843895a885a33bf16f251f85106/lib/truemail/wrapper.rb#L19

To do it just add into your Rails initializers `truemail.rb` file:

```ruby
# frozen_string_literal: true

module Truemail
  class Wrapper
    def call(&block)
      ::Timeout.timeout(timeout, &block)
    rescue ::Resolv::ResolvError, ::IPAddr::InvalidAddressError => error
      ::Logger.new($stdout).add(::Logger::ERROR) { error }
      false
    rescue ::Timeout::Error => error
      retry unless (self.attempts -= 1).zero?
      ::Logger.new($stdout).add(::Logger::ERROR) { error }
      false
    end
  end
end
```

Now we can see error context during network interactions... Please let me know about your results. Thanks!!!

**Comment by carlobeltrame** [comment]
Thanks, we'll try that and report back!

**Comment by kronn** [comment]
@bestwebua So, we just had the problem again, this time with logging you suggested. From what I see, Truemail runs into timeouts when requesting (in that order):

- mx_records
- cname_record
- a_record

After that, it concludes with a failure, because (understandably) `mx: target host(s) not found`. After restarting puma inside the container by touching `tmp/restart.txt`, the check runs fine again and the runtime for the check dropped from 12_000ms again to 11ms, which seems normal.

For the full backtrace, I attached a log-excerpt:

[truemail-dns-lookup-problem.log](https://github.com/truemail-rb/truemail/files/10199945/truemail-dns-lookup-problem.log)

I removed tokens and requests to our normal health-check, which only says "yeah, the rails-process is here". Otherwise, I left the log untouched.

**Comment by bestwebua** [comment]
@kronn Thanks. I'm 100 percent sure it's a network problem. Seems like to ensure that is it an network error you need add to ::Timeout::Error rescue block some network interaction.

**Comment by kronn** [comment]
I am a bit unsure, what you mean. 

Do you suggest adding some other network interaction in the rescue-block we added above?
That would be here:

https://github.com/hitobito/hitobito/blob/078f56e9861c81d28d9ed54f856e87694b7b6cf8/config/initializers/truemail_config.rb#L19-L23

Or did you mean something else? If so, please clarify :smiley:

**Comment by bestwebua** [comment]
@kronn I mean to add some kind of healthcheck of your system (some interaction with external host). Something like from example below:

```ruby
# frozen_string_literal: true

module Truemail
  class Wrapper
    def call(&block)
      ::Timeout.timeout(timeout, &block)
    rescue ::Resolv::ResolvError, ::IPAddr::InvalidAddressError => error
      ::Logger.new($stdout).add(::Logger::ERROR) { error }
      false
    rescue ::Timeout::Error => error
      retry unless (self.attempts -= 1).zero?
      ::Logger.new($stdout).add(::Logger::ERROR) { "Healthcheck: #{::Net::HTTP.get_response(URI('https://github.com')).code}" }
      ::Logger.new($stdout).add(::Logger::ERROR) { error }
      false
    rescue => error
      ::Logger.new($stdout).add(::Logger::ERROR) { "Network error: #{error}" }
      false
    end
  end
end
```

**Comment by kronn** [comment]
We decided to not debug this any further. If we did, we would try to determine the right cache of the network topology in truemail, ruby-stdlib, ruby or whereever. While interesting, this is a rather long process with little benefit.

We now check a known-good address with the known-good truemail and if this does not work, we restart the container. After that, everything works again. Since this is a Kubernetes-Feature, the problem is detected quickly and handled automatically.

Thanks for the debugging help.
